### ISSUE TITLE
problems with bibigrid-image

### ISSUE THREAD

**Issue by karl616** [BiBiServ/bibigrid#51]
Hi,

I'm trying to setup bibigrid and have problems creating the images. I found bibigrid-image on the linked page. Sadly it doesn't work. It won't get past fetching the keys. I think it is a problem of the firewall as the first key is successfully received. The second fails, but if I modify the keyserver flag in the displayed command (keyserver.ubuntu.com ==> hkp://keyserver.ubuntu.com:80) I can download the key manually. I have limited control over the firewall.

(If this is too much off-topic for this repository, I would appreciate a redirection)

**Comment by karl616** [comment]
I was able to work around the firewall. It would still be nice to fix that issue though.

Now the installation breaks trying to configure gridengine-master (Error message pasted below). Could it be that hostname has to follow a certain pattern? I can see in the code that bibigrid-master-$clusterId is used for the master. 



```
Installation returned the following message:
 -------
 | Preconfiguring packages ...
 | Selecting previously unselected package gridengine-master.
 | (Reading database ... 72498 files and directories currently installed.)
 | Preparing to unpack .../gridengine-master_6.2u5-7.3_amd64.deb ...
 | Unpacking gridengine-master (6.2u5-7.3) ...
 | Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
 | Processing triggers for ureadahead (0.100.0-16) ...
 | Setting up gridengine-master (6.2u5-7.3) ...
 | Initializing cluster with the following parameters:
 |  => SGE_ROOT: /var/lib/gridengine
 |  => SGE_CELL: default
 |  => Spool directory: /var/spool/gridengine/spooldb
 |  => Initial manager user: sgeadmin
 | Initializing spool (/var/spool/gridengine/spooldb)
 | Initializing global configuration based on /usr/share/gridengine/default-configuration
 | sge_gethostbyname failed
 | dpkg: error processing package gridengine-master (--configure):
 |  subprocess installed post-installation script returned error exit status 1
 | Processing triggers for ureadahead (0.100.0-16) ...
 | Errors were encountered while processing:
 |  gridengine-master
 | E: Sub-process /usr/bin/dpkg returned an error code (1)
 | 
 -------
```

**Comment by jkrue** [comment]
I need some more information to determine what's going wrong in your case.

Which Cloud environment do use  ? Openstack (which version), AWS, something else ?
Which Cloud-image do you use ?

However I will publish the sources of the latest bibigrid-image asap on GitHub. The linked binary is outdated and did not work very well behind a proxy/firewall.

**Comment by karl616** [comment]
Thank you very much.

The target environment is OpenStack (RedHat version 11, based on Ocata). 

That said, to get around the firewall I'm running everything locally through KVM. I'm new to this, but was able to set up a small config-driver for cloud-init with hostname defined to bibigrid-master. I am able to get through the bibigrid-image script if I do a manual installation of gridengine-master, autoconfiguring with master node: bibigrid-master. Still, the sge_gethostbyname errors return in Step 3, the GridEngine setup.

I'm using this image as base:

https://cloud-images.ubuntu.com/releases/14.04/release/ubuntu-14.04-server-cloudimg-amd64-disk1.img

When the new version is up, I will try it instead. And perhaps I'll try to setup the slave until then. I'm not sure, but maybe it doesn't need the gridengine-master package.

**Comment by karl616** [comment]
Hi again,

don't know if it is helpful, but I have found work-arounds allowing me to run the old bibigrid-image script without error. In the end it came down to two network related issues:

1) firewall blocking port 11371. This stops the import of repository keys

2) hostname issues stopping GridEngine to properly look up IP and hostnames

Solution to 1) As I couldn't change the code I wrapped gpg in a script that added hkp://*:80 to the keyserver names

Solution to 2) Adding the two following lines to /etc/hosts

10.0.0.12 bibigrid-master
127.0.0.1 bibigrid-master

Would you know if the DNS-server normally distributes this in ubuntu-based clouds?

**Comment by jkrue** [comment]
Hi, sorry for the late response. 

It takes me some time to move the bibigrid-image from our internal mercurial repository to GitHub (clean up, remove sensitive data, documentation, ...). However, since yesterday a [BibiGrid-Image](https://github.com/BiBiServ/bibigrid-images) repository exists on GitHub.

Comment to 2 (from BiBiGrid-Image known issues):
The default DNS entry (following the Openstack installation guide) for every instance has the pattern host-$ip(1st octet)-$ip(2nd octet)-$ip(3rd octect)-$ip(4th octet) independend of the given instance name. The cloud-init software, which typically configures an instance using the metadata service during boot time, hardcode the hostname to instance name (setting /etc/hosts). A bug (which only occurs in this context) in the dhcp-client software (since Ubuntu 14.04) prevents that DNS informations distributed by DHCP are updated when the instance has network connection. However, this problem can be solved. You can configure Openstack to use the instance-name as hostname (instead of above mentioned pattern) or patch the DHCP-client. The shell script **create_image.sh** located in scripts/openstack patch the standard ubuntu cloud images and  run bibigrid-image.

**Comment by karl616** [comment]
I'm just happy to get help :)

I will try this, but it sounds promising. I saw the bibigrid-image repository yesterday by coincidence, but did not figure understand how to solve the DNS problem. Thanks

**Comment by jkrue** [comment]
Related to the DNS issue - could you please try  the following:
1. Start an instance with an standard (non patched) ubuntu cloud image (14.04 or 16.04) should not make any difference.
2. Login into the instance 
3. What returns  _hostname_ ? 
4. Try to resolve the returned hostname using nslookup, dig , ... . Could it be resolved ?

**Comment by karl616** [comment]
I loaded an instance of the image linked in the openstack create_image.sh script.

hostname: ubuntu-bare-host

This is the same as the instance name I gave. This was currently done through the dashboard. (The behavior is the same if I use the command line client.) And no, it cannot be resolved. The configured nameserver is one from my cloud-provider, but it doesn't seem to resolve cloud-internal names.

After some searches, I understood that the openstack internal DNS is Designate. Browsing the available services (openstack catalog list) I cannot find it. 

I see three possibilities for me:
1. Convince my cloud-provider to provide DNSaas, Designate
2. Setup my own DNS in the subnet I intend to run bibigrid in and configure my instances to use this 
3. Have a DNS server running on the master node. At my current level I don't know how to make this flexible. Perhaps I learn through going through step 2...

I'm probably missing something

**Comment by jkrue** [comment]
That sounds weird. 
Could you resolve the hostname Openstack used internal for your VM ? If your local ip-address ist 192.168.0.100 then openstack internal uses "host-192-168.0.100" as hostname for this instance. 
A default setup should resolve at least the "internal hostname"

**Comment by karl616** [comment]
This didn't work on my first machine. I also tried host-192-168-0-100. I got my network setup with my account and to rule out this factor, I configured a new network and this seems to be a step in the right direction.

hostname: ubuntu-second-host
dig ubuntu-second-host: not resolved
dig host-192-168-0-14: resolved!

I will try to do the setup within this new network instead. I suspect that it will work.

**Comment by jkrue** [comment]
You have to use the patched images generated by the create-image.sh to get it work. As I mentioned before, a bug in the ubuntu cloud image dhcp-client prevents that a once set hostname can be changed.  Cloud init set the hostname before having network connection with the instacne name you set, the host-X-X-X hostnames are distributed via dhcp when network is available to the instance and should overwrite the cloud-init settings (which will not work because of the bug). Let me know if this will work ...

**Comment by karl616** [comment]
I was able to create the images and I think I figured out the problem. There are two levels to resolve:

1. internal, subnet level and hostnames
2. external, openstack servers, NFS-shares, and Internet

In my cloud, or at least as far as I have it figured out, these two levels are exclusive. The old network worked with a level-2 DNS server ==> no hostname resolving. Removing this, I cannot resolve anything outside my subnet. This sounds like something I have to solve myself.

Still, there is one more problem in this line; starting the cluster, the cloud-init configuration takes over again and hostnames default to instance names instead of DNS names. I double checked and it is the patched version of dhclient, but I guess it was never restarted after getting net access. Could that be?

**Comment by jkrue** [comment]
The DNS configuration of your OS cloud sounds very special. Maybe you have to ask your cloud administrators how to solve this. From my point of view there is no technical reason why the Openstack subnet DNS-Server should not resolve the subnet-internal hostnames and asked for anything else another external DNS Server.  This is the "default" way how DNS resolving works within Openstack (and maybe other clouds). When creating a (new) subnet you can of course add your own DNS Server for this subnet ignoring/replacing  any default settings.

In a typically OS configuration the DNS information is provided  by DHCP. In that case the (patched) version of the dhclient updated this information with the DCHP lease. In our cloud configuration the dhcp lease time is 120 seconds (which is OS default I thought), so every 2 minutes this information is updated. If no DNS information is provided by  DHCP (which can be configured) the patch has no effect.  If the  lease time is  greater than you have to wait longer.  If i would guess, I would say in your case this depends on your special OS configuration.

**Comment by karl616** [comment]
Yes, I start to understand that a lot of this is due to my local openstack setup and to some degree with my inexperience, why I appreciate the patience. The problem seems to be that the name servers doesn't escalate the queries, but simply refuse to handle them. As you suggest, this is something I have to figure out together with my cloud-provider.

According to /var/lib/dhcp/dhclient.*leases I seem to have a lease time of 24h. If the patch only kicks in on renewal I have to wait 12h to setup bibigrid. I submitted a work-around to the bibigrid repository, forcing the renewal to be part of the start-up procedure. It works, but I think that the slaves should be configured somehow else. As is I'm not sure if you want it in the main code, but it will work for me until I escalated the DNS issue.

Thank you for your help, and on my behalf, we could close this issue.
